### ISSUE TITLE
Ability to use company-mode selection for flyspell

### ISSUE THREAD

**Issue by dcostaras** [company-mode/company-mode#203]
Hi,

I'm trying to use company-mode for flyspell selection as shown here with popup.el in heading [Flyspell with popup.el](http://www.emacswiki.org/emacs/FlySpell). My elisp leaves a lot to be desired so I'm not sure if this is even possible.

What allows it to work is `popup-menu*` blocks while the selection is being made and then returns the selection made into the calling function (in this case `flyspell-emacs-popup`) for further use (particularly, in my case, adding the correction to an abbrev table).

Whereas `company-begin-backend` does not block and immediately returns a list of all candidates and then once the selection is made inserts the selection into the buffer. 

Is there a way to use the `company-psuedo-tooltip-*` functions in an analogous way to popup.el? Company handles overlays right and it doesn't look like popup.el will be ever fixed.

Much thanks,
Donavan

**Comment by Henry** [comment]
I have written a `company-flyspell` backend based on the backend I wrote for `completion-ui` but there is an issue which I had to hack around: some candidates might be shorter than the selected word which causes `company-pseudo-tooltip-unless-just-one-frontend` to fail.  To avoid this failure I put in code to convert the list of candidates to a list of strings and pad them with trailing spaces so that they are at least as long as the selected word:

         ;; Pad words to be at least as long and the length of `arg' to avoid
         ;; company-pseudo-tooltip-unless-just-one-frontend error
         ;;     "Args out of range"
         (progn
           (mapcar
            (lambda (str)
              (format (concat "%-" (number-to-string (length arg)) "s") str))
            words)))))

the downside of which is that the padding spaces are also inserted into the buffer when a short candidate is chosen.  Is there a better way than this in `company-mode`?  Are candidates shorter than the selected string currently supported?  If not could support be added?

**Comment by dgutov** [comment]
> some candidates might be shorter than the selected word which causes company-pseudo-tooltip-unless-just-one-frontend to fail

From a quick experiment, I don't see that. Could you show me the backtrace?

Make sure you're running Emacs 24.4 or newer, `M-x toggle-debug-on-error` and `M-x company-complete-common`.

**Comment by dgutov** [comment]
Here's the backend I tried:

```el
(defun company-aaa (command &optional value &rest args)
  (pcase command
    (`prefix (company-grab-symbol))
    (`candidates '("a" "bb" "ccc"))))
```

**Comment by Henry** [comment]
Thanks for the quick response.  I tried the your code snippet and it worked fine and reworked the company-flyspell backend and it also works fine; I don't know what I did wrong first time round.

Anyway here is the code for you to play with.  If you are happy with it feel free to include it as part of company-mode or if you think it needs further work please let me know.

https://github.com/Henry/dot-emacs/blob/master/my-lisp/company-flyspell.el

**Comment by dgutov** [comment]
The code looks very nice, and I'm happy to hear it works.

However, I'd much rather keep it as a separate package: it doesn't sound like a good candidate for the default value of `company-backends`, because, IIUC, the user will need to invoke it explicitly instead of using like a normal backend. Further, it will need some instructions to that effect, which you could put in README.

If you have the FSF copyright assignment signed, I'd be happy to accept it into the company-mode organization, while keeping your access, and also assist with publishing it to GNU ELPA.

**Comment by Henry** [comment]
I agree that I would not put it in the default list of backends but I think it is at least as useful as the ispell backend and for some probably more useful.  I have used it for years in `completion-ui` as my standard way of correcting spelling mistakes for which the `ispell` lookup approach is not very useful because the selected word must be a correctly spelled word beginning.

If `flyspell` is already running and `company-mode` installed and configured then `company-flyspell` just works either as a command, bound to a key or added to `company-backends`.  However I need to add some documentation to explain the difference to `company-ispell` which I think is not accurately named is it is actually using `lookup` and in fact `company-flyspell` is actually using `ispell`;  I think it would be good if this is also explained in `company-ispell.el`.  I can reproduce the header documentation in a README file if that would be useful.

I have contributed to emacs in the past and signed the FSF copyright assignment and would be happy to see this backend published in GNU ELPA.  Let me know what needs to be done.

Thanks for your help and responsiveness.

**Comment by dgutov** [comment]
>  However I need to add some documentation to explain the difference to company-ispell which I think is not accurately named is it is actually using lookup and in fact company-flyspell is actually using ispell;

Sure, pull request welcome.

> Let me know what needs to be done.

I've sent you an invite to join the org. After you accept, you should be able to create a new repo under it (let's call it the same name as the backend).

**Comment by stardiviner** [comment]
There is one thing to improve after I tested it out.
When the point is inside of word like `flysp|ell`, then `company-spell` complains it can't complete.
Should it be smart on decide current work or symbol, not just prefix?

**Comment by dgutov** [comment]
@stardiviner This won't work properly without a respective change in company-mode.

**Comment by stardiviner** [comment]
@dgutov I see, Thanks. `flyspell-popup` has some similar function, but it use `popup`.

**Comment by eflanigan00** [comment]
Any updates on this?
I would like to have the company numbers pop up when the flyspell corrections are listed.

**Comment by dgutov** [comment]
None. Also, since this discussion happene, and no further progress occured, I managed to forget why I invited @Henry and revoked his membership. :confused:
